File: apps/web/src/lib/deezer.ts
/**
 * Deezer search — free, no auth. One search call per track, plus one track
 * lookup for the release year (used by the reminiscence-bump analysis).
 * Yields album, preview, popularity rank, release year and a match score.
 */
import { getSql } from "./db";
import { getJson } from "./http";

const API = "https://api.deezer.com";

export interface DeezerMatch {
  deezerId: number | null;
  /** Deezer's artist ID — shared across every track by that artist regardless
   *  of how the source platform spelled the name. Powers KO↔EN auto-dedup. */
  artistId: number | null;
  /** Deezer's canonical artist name (e.g. "BTS" even if we searched "방탄소년단"). */
  artistName: string | null;
  album: string | null;
  coverUrl: string | null;
  previewUrl: string | null;
  releaseYear: number | null;
  rank: number | null; // Deezer popularity (higher = more mainstream)
  matchConfidence: number;
}

const EMPTY: DeezerMatch = {
  deezerId: null,
  artistId: null,
  artistName: null,
  album: null,
  coverUrl: null,
  previewUrl: null,
  releaseYear: null,
  rank: null,
  matchConfidence: 0,
};

/** Normalize for matching/scoring: lowercase, strip brackets/feat/symbols (keep CJK).
 *  Exported for unit testing (R37) — also used internally. */
export function norm(s: string): string {
  return s
    .toLowerCase()
    .replace(/\([^)]*\)|\[[^\]]*\]/g, " ")
    .replace(/\bfeat\.?.*$/i, " ")
    .replace(/[^a-z0-9가-힣ぁ-んァ-ヶ一-龯]+/gi, " ")
    .trim();
}

/** Title/artist similarity 0.4–0.95. Exported for unit testing (R37). */
export function scoreMatch(a: string, b: string): number {
  const x = norm(a);
  const y = norm(b);
  if (!x || !y) return 0.4;
  if (x === y) return 0.95;
  if (x.includes(y) || y.includes(x)) return 0.75;
  return 0.5;
}

/** Parses a 4-digit year out of a Deezer release_date ("YYYY-MM-DD"). */
function yearOf(date: unknown): number | null {
  const m = typeof date === "string" ? date.match(/^(\d{4})/) : null;
  if (!m) return null;
  const y = Number(m[1]);
  return y >= 1900 && y <= new Date().getFullYear() + 1 ? y : null;
}

export async function searchDeezer(
  artist: string,
  title: string,
  opts: { withYear?: boolean } = {},
): Promise<DeezerMatch> {
  const withYear = opts.withYear ?? true;
  const cacheKey = `${artist.toLowerCase().trim()}|${title.toLowerCase().trim()}`;
  const sql = getSql();

  // Cached match — avoids a repeat Deezer call (faster, fewer rate limits).
  try {
    const hitRow = await sql`SELECT payload FROM deezer_match WHERE cache_key = ${cacheKey}`;
    if (hitRow.length > 0) return hitRow[0].payload as DeezerMatch;
  } catch {
    /* fall through to a live lookup */
  }

  const cleanTitle = norm(title) || title;
  const advanced = `artist:"${artist.replace(/"/g, "")}" track:"${cleanTitle}"`;

  let data = await getJson(`${API}/search?q=${encodeURIComponent(advanced)}&limit=1`);
  let hit = data?.data?.[0];
  if (!hit) {
    data = await getJson(`${API}/search?q=${encodeURIComponent(`${artist} ${title}`)}&limit=1`);
    hit = data?.data?.[0];
  }
  if (!hit) return EMPTY;

  // The search result carries rank but not a release date — one extra
  // lookup gets the original year (skipped when the caller doesn't need it).
  let releaseYear: number | null = null;
  if (withYear && hit.id) {
    try {
      const track = await getJson(`${API}/track/${hit.id}`);
      releaseYear = yearOf(track?.release_date) ?? yearOf(track?.album?.release_date);
    } catch {
      /* year stays null */
    }
  }

  // Confidence has to weight BOTH title AND artist. The previous
  // title-only score was unsafe for music: many songs share titles
  // (covers, identical-name tracks across decades), and the YouTube
  // Music artist string we send in can drift from Deezer's canonical
  // ("BTS" vs "방탄소년단" vs "BTS (방탄소년단)" vs "Topic" channels).
  // Title carries slightly more weight than artist because translit /
  // alias mismatches on artist are common even for genuine matches.
  // Album would tighten this further but isn't in the search-input
  // contract today; threaded in a follow-up.
  const titleScore = scoreMatch(title, (hit.title as string | undefined) ?? "");
  const artistScore = scoreMatch(
    artist,
    (hit.artist?.name as string | undefined) ?? "",
  );
  const matchConfidence = titleScore * 0.55 + artistScore * 0.45;

  // Below this floor we don't trust the match enough to surface its
  // popularity / release year / preview as if they describe the user's
  // actual track. Keep the basic identifiers (deezerId, artist, album,
  // cover) since those just locate the candidate — but suppress the
  // year + rank + preview, which downstream features read as ground
  // truth (reminiscence-bump uses year, recommend popularity uses rank,
  // preview powers the play button). A bad year would silently shift
  // the listener's "nostalgia window" to the wrong decade.
  const HIGH_CONFIDENCE = 0.65;
  const isStrongMatch = matchConfidence >= HIGH_CONFIDENCE;

  const result: DeezerMatch = {
    deezerId: hit.id ?? null,
    artistId: typeof hit.artist?.id === "number" ? hit.artist.id : null,
    artistName: typeof hit.artist?.name === "string" ? hit.artist.name : null,
    album: hit.album?.title ?? null,
    coverUrl: hit.album?.cover_big || hit.album?.cover_medium || null,
    previewUrl: isStrongMatch ? hit.preview || null : null,
    releaseYear: isStrongMatch ? releaseYear : null,
    rank: isStrongMatch && typeof hit.rank === "number" ? hit.rank : null,
    matchConfidence,
  };
  // Cache both hits and misses. Without caching the EMPTY result, tracks
  // Deezer can't match get re-searched on every batch — wasting subrequests
  // and (on Workers free plan) tripping the 50-subreq cap on the same 14
  // dead-end tracks forever.
  try {
    await sql`
      INSERT INTO deezer_match (cache_key, payload)
      VALUES (${cacheKey}, ${JSON.stringify(result)}::jsonb)
      ON CONFLICT (cache_key) DO NOTHING`;
  } catch {
    /* caching is best-effort */
  }
  return result;
}

File: services/analysis/app/pipeline/deezer.py
"""Deezer search — Python port of apps/web/src/lib/deezer.ts.

Behaviour is intentionally identical: same cache key, same scoring, same
fallback chain. The `deezer_match` table is shared between this service
and the TS code, so a hit written by either runner serves both.

The single semantic addition over the TS version is that misses (artists
Deezer can't find) are also cached — without that, the same dead-end
tracks were getting re-searched on every batch.
"""

from __future__ import annotations

import asyncio
import json
import re
import urllib.parse
from typing import Any

import asyncpg
import httpx

API = "https://api.deezer.com"

# Global cap on simultaneous outbound Deezer requests across all users.
# With multi-user parallel worker ticks (asyncio.gather) and 30 parallel
# enrichTrack calls per user, unbounded concurrency would burst into 300+
# requests and trip Deezer's rate limits. ~25 keeps us comfortably under
# their unwritten ceiling while still finishing a batch in seconds.
_HTTP_SEMA = asyncio.Semaphore(25)

# Strip CJK-friendly: lowercase, drop bracketed bits, drop "feat. ...",
# collapse non-alphanumeric (keeping Hangul / Hiragana / Katakana / Han).
_norm_strip = re.compile(r"\([^)]*\)|\[[^\]]*\]")
_norm_feat = re.compile(r"\bfeat\.?.*$", re.IGNORECASE)
_norm_punct = re.compile(r"[^a-z0-9가-힣ぁ-んァ-ヶ一-龯]+", re.IGNORECASE)


def _norm(s: str) -> str:
    s = s.lower()
    s = _norm_strip.sub(" ", s)
    s = _norm_feat.sub(" ", s)
    s = _norm_punct.sub(" ", s)
    return s.strip()


def _score(a: str, b: str) -> float:
    x, y = _norm(a), _norm(b)
    if not x or not y:
        return 0.4
    if x == y:
        return 0.95
    if x in y or y in x:
        return 0.75
    return 0.5


def _year_of(date: Any) -> int | None:
    if not isinstance(date, str):
        return None
    m = re.match(r"^(\d{4})", date)
    if not m:
        return None
    y = int(m.group(1))
    # Same sanity window the TS uses — pre-1900 / far-future are garbage.
    import datetime as _dt
    if 1900 <= y <= _dt.date.today().year + 1:
        return y
    return None


EMPTY: dict[str, Any] = {
    "deezerId": None,
    "artistId": None,
    "artistName": None,
    "album": None,
    "coverUrl": None,
    "previewUrl": None,
    "releaseYear": None,
    "rank": None,
    "matchConfidence": 0,
}


class Deezer:
    """Async Deezer client with Neon-backed caching + the same scoring as TS."""

    def __init__(self, http: httpx.AsyncClient, pool: asyncpg.Pool) -> None:
        self.http = http
        self.pool = pool

    async def search(
        self, artist: str, title: str, *, with_year: bool = True,
    ) -> dict[str, Any]:
        cache_key = f"{artist.lower().strip()}|{title.lower().strip()}"

        # 1. Cache lookup — both hits and misses are cached now, so this is
        #    the fast path for ~95% of tracks once the service has been
        #    running for a while.
        async with self.pool.acquire() as conn:
            try:
                row = await conn.fetchrow(
                    "SELECT payload FROM deezer_match WHERE cache_key = $1",
                    cache_key,
                )
                if row is not None:
                    payload = row["payload"]
                    if isinstance(payload, str):
                        payload = json.loads(payload)
                    return payload
            except Exception:
                pass  # cache outage → live lookup

        # 2. Advanced query first (artist + track field separators give
        #    Deezer the best chance to disambiguate).
        clean_title = _norm(title) or title
        advanced = f'artist:"{artist.replace(chr(34), "")}" track:"{clean_title}"'
        data = await self._search_raw(advanced)
        hit = (data.get("data") or [None])[0] if data else None

        if not hit:
            data = await self._search_raw(f"{artist} {title}")
            hit = (data.get("data") or [None])[0] if data else None

        if not hit:
            await self._cache(cache_key, EMPTY)
            return EMPTY

        release_year: int | None = None
        if with_year and hit.get("id"):
            try:
                track = await self._get_json(f"{API}/track/{hit['id']}")
                release_year = _year_of((track or {}).get("release_date")) or _year_of(
                    ((track or {}).get("album") or {}).get("release_date"),
                )
            except Exception:
                pass  # year stays None

        artist_obj = hit.get("artist") or {}
        album_obj = hit.get("album") or {}
        result = {
            "deezerId": hit.get("id"),
            "artistId": artist_obj.get("id") if isinstance(artist_obj.get("id"), int) else None,
            "artistName": artist_obj.get("name") if isinstance(artist_obj.get("name"), str) else None,
            "album": album_obj.get("title"),
            "coverUrl": album_obj.get("cover_big") or album_obj.get("cover_medium"),
            "previewUrl": hit.get("preview") or None,
            "releaseYear": release_year,
            "rank": hit["rank"] if isinstance(hit.get("rank"), int) else None,
            "matchConfidence": _score(title, hit.get("title") or ""),
        }
        await self._cache(cache_key, result)
        return result

    async def _search_raw(self, q: str) -> dict[str, Any] | None:
        return await self._get_json(
            f"{API}/search?q={urllib.parse.quote(q)}&limit=1",
        )

    async def _get_json(self, url: str, timeout: float = 6.0) -> dict[str, Any] | None:
        """Fetch JSON, returning None on any failure or top-level `error`.

        Guarded by the module-level `_HTTP_SEMA` so the global Deezer call
        rate stays bounded even under high user concurrency."""
        try:
            async with _HTTP_SEMA:
                r = await self.http.get(url, timeout=timeout)
            if r.status_code >= 400:
                return None
            data = r.json()
            if isinstance(data, dict) and data.get("error"):
                return None
            return data
        except Exception:
            return None

    async def _cache(self, key: str, payload: dict[str, Any]) -> None:
        async with self.pool.acquire() as conn:
            try:
                await conn.execute(
                    """
                    INSERT INTO deezer_match (cache_key, payload)
                    VALUES ($1, $2::jsonb)
                    ON CONFLICT (cache_key) DO NOTHING
                    """,
                    key,
                    json.dumps(payload),
                )
            except Exception:
                pass  # caching is best-effort
